File: database/create_pada_csv.pl
#!/usr/bin/perl
use strict;
use warnings;
use open ':std', ':encoding(UTF-8)';

open(my $data, '<', "database/amara_pada.utf8") or die "Could not open input file $!\n";
open(my $output, '>', "database/tmp_pada.csv") or die "Could not open output file $1\n";

my %mula_artha = (
    # flintstones    => [ "fred", "barney" ],
    # jetsons        => [ "george", "jane", "elroy" ],
    # simpsons       => [ "homer", "marge", "bart" ],
);

my $uid=1;

while (my $line = <$data>) {
  chomp $line;

  $line =~ s/_/ /g;
  my @fields = split "," , $line;

  # The identifier in $fields[0] is a permanent unique id for coordinating
  # translations etc. We will generate a new identifier based on the file order
  # my $uid = $fields[0];

  my $pada_uid = $fields[0];
  my $pada = $fields[1];
  my $varga_number = $fields[2] =~ s/^(.*\..*)\..*\..*\..*$/$1/r;
  my $sloka_number = $fields[2] =~ s/^(.*\..*\..*)\..*\..*$/$1/r;
  my $sloka_line = $fields[2] =~ s/^(.*\..*\..*\..*)\..*$/$1/r;
  my $sloka_word = $fields[2];
  my $linga = $fields[3];
  my $varga = $fields[4];
  my $artha = $fields[5];

  if (! $mula_artha{$sloka_number}) {
    $mula_artha{$sloka_number} = [];
  }

  push $mula_artha{$sloka_number}, $artha;

  printf $output ("%d,%d,%s,%s,%s,%s,%s,%s,%s,,%s\n", $uid++, $pada_uid, $varga_number, $sloka_number, $sloka_line, $sloka_word, $pada, $linga, $varga, $artha);

  # printf $output ("%d,%s,%s,%s,%s,%s,%s\n", $uid++, $fields[0], $sloka_number, $fields[1], $fields[2], $fields[3], $fields[4]);
}

# foreach my $item (@{$HoA{$key}})
# {
#     print "Test item: $nextItem\n";
# }

File: database/create_pada_csv.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-

import io, re
from indic_transliteration import sanscript, xsanscript
from indic_transliteration.sanscript import SchemeMap, SCHEMES, transliterate


uid=1

with open('database/amara_pada.csv') as datafile:
    with open('database/tmp_pada.csv', 'w') as outputfile:

        while True:
            line = datafile.readline()
            if not line:
                break

            line = line.replace('_', ' ')
            parts = line.split(',')

          # The identifier in $fields[0] is a permanent unique id for coordinating
          # translations etc. We will generate a new identifier based on the file order
          # my $uid = $fields[0];

            pada_uid = parts[0]
            pada = parts[1]
            pada_slp1 = transliterate(pada, xsanscript.DEVANAGARI, xsanscript.SLP1)

            sloka_word = parts[2]
            linga = parts[3].replace('.', '')
            varga = parts[4]
            artha = parts[5]

            re_parts = re.search (r'^([^.]+)\.([^.]+)\.([^.]+)\.([^.]+)\.([^.]+)', sloka_word);
            re_parts_variant = re.search (r'^([^.]+)\.([^.]+)\.([^.]+)\.([^.]+)\.([^.]+)\.([^.]+)', sloka_word);
            if re_parts:
                varga_number = "%s.%s" % (re_parts.group(1), re_parts.group(2))
                sloka_number = "%s.%s.%s" % (re_parts.group(1), re_parts.group(2), re_parts.group(3))
                sloka_line = "%s.%s.%s.%s" % (re_parts.group(1), re_parts.group(2), re_parts.group(3), re_parts.group(4))
                variant_main_sloka_word = "%s.%s.%s.%s.%s" % (re_parts.group(1), re_parts.group(2), re_parts.group(3), re_parts.group(4), re_parts.group(5))
                is_variant = 0

            if re_parts_variant:
                is_variant = 1
            else:
                variant_main_sloka_word = ''

            outputfile.write("%d,%s,%s,%s,%s,%s,%d,%s,%s,%s,%s,%s,,,%s\n" % (uid, pada_uid, varga_number, sloka_number, sloka_line, sloka_word, is_variant, variant_main_sloka_word, pada, pada_slp1, linga, varga, artha))
            uid = uid+1
